src/audio/rtnr/rtnr.c

```diff
@@ -450,16 +424,25 @@ static int32_t rtnr_set_value(struct comp_dev *dev, struct sof_ipc_ctrl_data *cd
 {
 	uint32_t val = 0;
 	int32_t j;
+	struct comp_data *cd = comp_get_drvdata(dev);
 
 	for (j = 0; j < cdata->num_elems; j++) {
 		val |= cdata->chanv[j].value;
-		comp_info(dev, "rtnr_set_value(), value = %u", val);
+		comp_dbg(dev, "rtnr_set_value(), value = %u", val);
 	}
 
 	if (val)
-		rtnr_set_process(dev);
+		cd->process_enable = true;
 	else
-		rtnr_set_bypass(dev);
+		cd->process_enable = false;
+
+	if (cd->process_enable) {
+		comp_info(dev, "rtnr_set_value(): enabled");
+		cd->rtnr_copy_func = rtnr_process;
+	} else {
+		comp_info(dev, "rtnr_set_value(): passthrough");
+		cd->rtnr_copy_func = rtnr_passthrough;
+	}
```

Comment:
Just as I mentioned in https://github.com/thesofproject/sof/pull/5616/files#r838305750
I am worried about that rtnr_copy_func may be out of sync. What would happen if rtnr_copy_func is changed while it is under processing?
Will need the comment from the expert.

Reply:
How about using single copy_func, but check the process_enable flag inside copy_func?

Reply:
Yes that would be better than the current code. rtnr_set_value() only do setting cd->process_enable value.
In rtnr_copy(), check cd->process_enable to determine the path. I think it could minimize the risk of asynchronous issue.

A quick question, when this is implemented, it means we can switch path in units of copy() (data sample chunk in one scheduled task). Would rtnr effect be degraded since it will obtain intermittent data samples, e.g. enabled during 1st to 100th task and then bypassed, re-enabled from 200th task. Fow now I could think of some factors which may drop the effect and even generate wrong output data.
1. is there sample latency in rtnr? That is, the processed data has a certain delay than the source data.
2. due to the discontinuity of data sample chunks, the waveform could be distorted if some algorithm uses a processing window larger than one chunk.

That's also the reason we tend to change path/config during the component restart (reset and prepare). In so doing the component actually regard each segment as an individual stream.

Reply:
Hi Johny,

There will be latency, but in overall it should be acceptable unless the user is toggling on&off rapidly.